### PR TITLE
[webview_flutter_wkwebview] The rest of the Objective-C HostApi methods 

### PR DESCRIPTION
**Pull request by bparrishMines**
This PR implements the rest of the `HostApi`s by following the pattern approved in https://github.com/flutter/plugins/pull/5341.

Also include minor fixes such as misspellings, using methods supported on all iOS versions, and annotating some methods with `@async`.

No version change:
Part of https://github.com/flutter/flutter/issues/93732 and doesn't make any changes to the current implementation.

No CHANGELOG change: Incremental unused code doesn't need to be noted in the CHANGELOG.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit_webview_widget.dart [hunk 1]
```diff
@@ -517,8 +517,9 @@ class WebKitWebViewPlatformController extends WebViewPlatformController {
     Set<String> removedJavaScriptChannels = const <String>{},
   }) async {
     webView.configuration.userContentController.removeAllUserScripts();
-    webView.configuration.userContentController
-        .removeAllScriptMessageHandlers();
+    _scriptMessageHandlers.keys.forEach(
+      webView.configuration.userContentController.removeScriptMessageHandler,
+    );
```
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit_webview_widget.dart:522]: `WKUserContentController.removeAllScriptMessageHandlers()` is only available 14+

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart [hunk 2]
```diff
@@ -101,19 +101,19 @@ class WKAudiovisualMediaTypeEnumData {
...
+  localStorage,
   sessionStorage,
-  sqlDatabases,
+  webSQLDatabases,
   indexedDBDatabases,
 }
 
-class WKWebsiteDataTypesEnumData {
-  late WKWebsiteDataTypesEnum? value;
+class WKWebsiteDataTypeEnumData {
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:115]: Can we eliminate this wrapper? I forget if direct used of null enums are part of what we fixed, or if they are still broken.

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart [hunk 3]
```diff
@@ -207,7 +207,11 @@ class WKScriptMessageData {
 ///
 /// See https://developer.apple.com/documentation/foundation/nshttpcookie?language=objc.
 class NSHttpCookieData {
-  late Map<NSHttpCookiePropertyKeyEnumData?, String?> properties;
+  // `NSDictionary`s are unable to use data classes as keys because they don't
+  // conform to `NSCopying`. This splits the map of properties into a list of
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:211]: Could you file an issue about this and link to it from the comment? It seems like something we could add support for.

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart [hunk 4]
```diff
@@ -227,12 +231,12 @@ abstract class WKWebsiteDataStoreHostApi {
...
   @ObjCSelector(
-    'removeDataFromDataStoreWithIdentifier:ofTypes:secondsModifiedSinceEpoch:',
+    'removeDataFromDataStoreWithIdentifier:ofTypes:modifiedSince:',
   )
   @async
   bool removeDataOfTypes(
     int instanceId,
-    List<WKWebsiteDataTypesEnumData> dataTypes,
+    List<WKWebsiteDataTypeEnumData> dataTypes,
     double secondsModifiedSinceEpoch,
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:240]: Nit: `modificationTimeInSecondsSinceEpoch`?

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit_webview_widget.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit_webview_widget.dart:522]: Maybe leave a TODO to use `removeAllScriptMessageHandlers` once we have a Dart-side equivalent to `@availability`?

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit_api_impls.dart [hunk 5]
```diff
@@ -10,52 +10,50 @@ import '../common/web_kit.pigeon.dart';
...
-            key.toNSHttpCookiePropertyKeyEnumData(),
-            value.toString(),
-          );
+      propertyKeys: properties.keys.map<NSHttpCookiePropertyKeyEnumData>(
+        (NSHttpCookiePropertyKey key) {
+          return key.toNSHttpCookiePropertyKeyEnumData();
         },
-      ),
+      ).toList(),
+      propertyValues: properties.values.map<Object>((_) => _).toList(),
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit_api_impls.dart:56]: It's not clear to me from the docs that Dart map orders for unordered implementations have to be consistent, so this makes me nervous.

I would rather you extract `properties.keys` to a local `Iterable`, and then use it as the driver for both map operations. It'll be a little less efficient, but guaranteed to be correct.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart [hunk 6]
```diff
@@ -13,6 +13,7 @@ import 'web_kit_api_impls.dart';
 /// Times at which to inject script content into a webpage.
 ///
 /// Wraps [WKUserScriptInjectionTime](https://developer.apple.com/documentation/webkit/wkuserscriptinjectiontime?language=objc).
+
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart:16]: Why all the whitespace between comments and the things they are commenting in this file?

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebsiteDataStoreHostApi.m [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+              [dataStore
+                  removeDataOfTypes:stringDataTypes
+                      modifiedSince:[NSDate dateWithTimeIntervalSince1970:secondsModifiedSinceEpoch
+                                                                              .doubleValue]
+                  completionHandler:^{
+                    if (records.count > 0) {
+                      completion(@YES, nil);
+                    } else {
+                      completion(@NO, nil);
+                    }
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebsiteDataStoreHostApi.m:67]: `completion(@(records.count > 0), nil);` ?

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFUserContentControllerHostApi.m [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                                                            error:
+                                                                (FlutterError *_Nullable *_Nonnull)
+                                                                    error {
+  if (@available(iOS 14.0, *)) {
+    [[self userContentControllerForIdentifier:instanceId] removeAllScriptMessageHandlers];
+  } else {
+    *error = [FlutterError
+        errorWithCode:@"FWFUnsupportedVersionError"
+              message:@"removeAllScriptMessageHandlers is only supported on versions 14+."
+              details:nil];
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFUserContentControllerHostApi.m:64]: Not blocking getting this new implementation completed and deployed, but we should audit the new webview implementations as a follow-up later to do the error mapping from `PlatformExceptions` that's described in the wiki now. E.g., we should have a specific, platform-neutral, package-specific Dart error+code for operations that aren't supported on the current platform.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.m [hunk 9]
```diff
@@ -23,3 +23,132 @@
...
+      return NSHTTPCookieOriginURL;
+    case FWFNSHttpCookiePropertyKeyEnumPath:
+      return NSHTTPCookiePath;
+    case FWFNSHttpCookiePropertyKeyEnumPort:
+      return NSHTTPCookiePort;
+    case FWFNSHttpCookiePropertyKeyEnumSameSitePolicy:
+      if (@available(iOS 13.0, *)) {
+        return NSHTTPCookieSameSitePolicy;
+      } else {
+        NSLog(@"NSHTTPCookieSameSitePolicy is only supported on iOS 13+.");
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.m:83]: We probably don't want to log about this; people aren't going to want to conditionalize their Dart cookie construction on iOS version. I vote we just silently drop it on unsupported versions.

**Review by stuartmorgan — APPROVED** [review]
LGTM with small nits!

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart (hunk 2 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:115]: This is still required due to: https://github.com/flutter/flutter/issues/87307. I updated all the wrappers with a `TODO`. However, it looks like https://github.com/flutter/flutter/issues/100594 was fixed. So, I updated all the data classes.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+                       error:(FlutterError *_Nullable *_Nonnull)error {
+  FWFNavigationDelegate *navigationDelegate = [[FWFNavigationDelegate alloc] init];
+  [self.instanceManager addInstance:navigationDelegate withIdentifier:instanceId.longValue];
+}
+
+- (void)setDidFinishNavigationForDelegateWithIdentifier:(nonnull NSNumber *)instanceId
+                                     functionIdentifier:(nullable NSNumber *)functionInstanceId
+                                                  error:(FlutterError *_Nullable __autoreleasing
+                                                             *_Nonnull)error {
+  // TODO(bparrishMines): Implement when callback method design is finalized.
```
Comment (cyanglaz) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m:38]: Do we have an issue tracking this?

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebViewConfigurationHostApi.m [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      switch (data.value) {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+        case FWFWKAudiovisualMediaTypeEnumNone:
+          configuration.requiresUserActionForMediaPlayback = false;
+          break;
+        case FWFWKAudiovisualMediaTypeEnumAudio:
+        case FWFWKAudiovisualMediaTypeEnumVideo:
+        case FWFWKAudiovisualMediaTypeEnumAll:
+          configuration.requiresUserActionForMediaPlayback = true;
```
Comment (cyanglaz) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebViewConfigurationHostApi.m:80]: nits:
add a break to the end of the case. 
Maybe also add an empty default case.

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart (hunk 3 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:227]: Done: https://github.com/flutter/flutter/issues/103383

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart (hunk 6 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart]: Accident! I used annotations and a template processor to generate most of the Objective-C code. I assumed the formatter would remove the Dart space once the annotations were removed.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.m (hunk 9 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.m]: Agreed. I updated the enum with documentation about the version requirement and that it is ignored if used.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit_api_impls.dart (hunk 5 above)
Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit_api_impls.dart]: Oops, I was not looking at the right part of the docs. That does indeed address my concern.

> I'm not sure what you mean by unordered implementations

I was talking about this part of the docs:
> The plain [HashMap](https://api.dart.dev/stable/2.17.0/dart-collection/HashMap-class.html) is unordered (no order is guaranteed),
the [LinkedHashMap](https://api.dart.dev/stable/2.17.0/dart-collection/LinkedHashMap-class.html) iterates in key insertion order,
and a sorted map like [SplayTreeMap](https://api.dart.dev/stable/2.17.0/dart-collection/SplayTreeMap-class.html) iterates the keys in sorted order.

My concern was that for `HashMap` there might not be any guarantees *at all* about order, including between calling `keys` and then calling `values` afterward. But that would violate the `values` docs you referenced.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m (hunk 10 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m:38]: This is just referring to the general rework PR: https://github.com/flutter/flutter/issues/93732. The followup to this PR is to implement this single callback method and to get feedback on the design.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebViewConfigurationHostApi.m (hunk 11 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebViewConfigurationHostApi.m:80]: Added the `break`, but I could only add an empty default case with a semicolon

```objc
default:;

//or

default:
  ;
```

**Pull request merged**